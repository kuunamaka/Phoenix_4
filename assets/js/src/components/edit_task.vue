<template>
  <table>
    <div>
      <h2>Updating Task</h2>
      <p v-if="errors.length">Please correct the following error(s):</p>

      <ul>
        <li v-for="error in errors" :key="error" class="error">{{ error }}</li>
      </ul>
      <div>
        Name
        <input v-model="name" />
      </div>
      <div>
        Assigned by:
        <select v-model="assignee_id">
          <option disabled>Please select one</option>
          <option>
            {{ task.assignee_id }}
          </option>
        </select>
      </div>
      <div>
        Has finished?
        <input type="checkbox" id="is_finished" v-model="is_finished" />
      </div>
      <button @click="updateTask(task)">Submit</button>
      <a href="/tasks" class="button">Exit</a>
    </div>
    <delete-task />
  </table>
</template>

<script>
import axios from "axios";
import DeleteTask from "./delete_task";
export default {
  name: "edit_task",
  components: { DeleteTask },
  data() {
    return {
      name: "",
      assignee_id: null,
      is_finished: false,
      task: "",
      errors: [],
    };
  },
  methods: {
    async updateTask(task) {
      this.errors = [];
      if (!this.name) {
        this.errors.push("Name required");
      }
      if (!this.assignee_id) {
        this.errors.push("Assignee required");
      }
      if (this.errors.length !== 0) {
        return;
      }
      await axios.put(`/api/tasks/${task.id}`, {
        task: {
          name: this.name,
          is_finished: this.is_finished,
          assignee_id: this.assignee_id,
        },
      });
      window.location.href = "/tasks";
    },
  },
  async mounted() {
    const task_id = window.location.pathname.split("/")[2];
    const resp = await axios.get(`/api/tasks/${task_id}`);
    this.task = resp.data;
  },
};
</script>
