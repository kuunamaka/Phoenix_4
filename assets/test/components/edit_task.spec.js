import { shallowMount } from "@vue/test-utils";
import edit_task from "@/src/components/edit_task.vue";

const axios = require("axios");
const MockAdapter = require("axios-mock-adapter");
const mock = new MockAdapter(axios);

mock.onGet("/api/tasks/1").reply(200, {
  id: 1,
  name: "Test",
  is_finished: false,
  assignee_id: 1,
});

global.window = Object.create(window);
const pathname = "/tasks/1";

Object.defineProperty(window, "location", {
  value: {
    pathname,
    href: `http://localhost:4001${pathname}`,
  },
});

describe("edit_task.vue", () => {
  const wrapper = shallowMount(edit_task);
  const vm = wrapper.vm;

  it("renders entire component correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  describe("updateTask()", () => {
    it("redirect to /tasks", async () => {
      mock.onPut("/api/tasks/1").reply(200, {
        task: {
          id: 1,
          name: "updateTest",
          is_finished: false,
          assignee_id: 1,
        },
      });
      await wrapper.setData({
        id: 1,
        name: "updateTest",
        is_finished: false,
        assignee_id: 1,
      });

      await vm.updateTask({
        id: 1,
        name: "updateTest",
        is_finished: false,
        assignee_id: 1,
      });

      expect(window.location.href).toEqual("/tasks");
    });

    it("show validate error messages", async () => {
      mock
        .onPut("/api/tasks/1", {
          task: { id: 1, name: "", is_finished: false, assignee_id: 1 },
        })
        .reply(404);
      await wrapper.setData({
        id: 1,
        name: "",
        is_finished: false,
        assignee_id: 1,
      });

      const updateButton = wrapper.find("button");
      await updateButton.trigger("click");

      expect(wrapper.find("p").element).toMatchSnapshot();
    });
  });
});
