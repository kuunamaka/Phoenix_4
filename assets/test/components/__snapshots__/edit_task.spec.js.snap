// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit_task.vue is a Vue instance 1`] = `
<table>
  <div>
    <h2>
      Updating Task
    </h2>
     
    <!---->
     
    <div>
      
      Name
      
      <input />
    </div>
     
    <div>
      
      Assigned by:
      
      <select>
        <option
          disabled="disabled"
        >
          Please select one
        </option>
         
        <option>
          
          1
        
        </option>
      </select>
    </div>
     
    <div>
      
      Has finished?
      
      <input
        id="is_finished"
        type="checkbox"
      />
    </div>
     
    <button>
      Submit
    </button>
     
    <a
      class="button"
      href="/tasks"
    >
      Exit
    </a>
  </div>
   
  <delete-task-stub />
</table>
`;

exports[`edit_task.vue updateTask() show validate error messages 1`] = `
<p>
  
      Please correct the following error(s):
      
  <ul>
    <li
      class="error"
    >
       Name required
    </li>
  </ul>
</p>
`;
